pkg/decision/rollout_service_test.go

```diff
@@ -89,27 +124,98 @@ func (s *RolloutServiceTestSuite) TestGetDecisionHappyPath() {
 	s.mockExperimentService.AssertExpectations(s.T())
 }
 
-func (s *RolloutServiceTestSuite) TestGetDecisionFailsBucketing() {
+func (s *RolloutServiceTestSuite) TestGetDecisionFallbacksToLastWhenFailsBucketing() {
 	// Test experiment passes targeting but not bucketing
-	testExperimentBucketerDecision := ExperimentDecision{
+	testExperiment1112BucketerDecision := ExperimentDecision{
 		Decision: Decision{
 			Reason: reasons.NotBucketedIntoVariation,
 		},
 	}
-
+	testExperiment1118BucketerDecision := ExperimentDecision{
+		Variation: &testExp1118Var2224,
+		Decision:  Decision{Reason: reasons.BucketedIntoVariation},
+	}
+	experiment1118DecisionContext := ExperimentDecisionContext{
+		Experiment:    &testExp1118,
+		ProjectConfig: s.mockConfig,
+	}
 	s.mockAudienceTreeEvaluator.On("Evaluate", testExp1112.AudienceConditionTree, s.testConditionTreeParams).Return(true, true)
-	s.mockExperimentService.On("GetDecision", s.testExperimentDecisionContext, s.testUserContext).Return(testExperimentBucketerDecision, nil)
+	s.mockAudienceTreeEvaluator.On("Evaluate", testExp1118.AudienceConditionTree, s.testConditionTreeParams).Return(true, true)
+	s.mockExperimentService.On("GetDecision", s.testExperiment1112DecisionContext, s.testUserContext).Return(testExperiment1112BucketerDecision, nil)
+	s.mockExperimentService.On("GetDecision", experiment1118DecisionContext, s.testUserContext).Return(testExperiment1118BucketerDecision, nil)
+
 	testRolloutService := RolloutService{
 		audienceTreeEvaluator:     s.mockAudienceTreeEvaluator,
 		experimentBucketerService: s.mockExperimentService,
-		logger:logging.GetLogger("sdkKey", "RolloutService"),
+		logger:                    logging.GetLogger("sdkKey", "RolloutService"),
 	}
 	expectedFeatureDecision := FeatureDecision{
+		Experiment: testExp1118,
+		Variation:  &testExp1118Var2224,
+		Source:     Rollout,
+		Decision:   Decision{Reason: reasons.BucketedIntoRollout},
+	}
+	decision, _ := testRolloutService.GetDecision(s.testFeatureDecisionContext, s.testUserContext)
+	s.Equal(expectedFeatureDecision, decision)
+	s.mockAudienceTreeEvaluator.AssertExpectations(s.T())
```

Comment:
Can we also assert, number of rolloutrules `experiments` were in it? and the returned mocked roleout rule was the last one. Also throw `exception` for all other roleout rules so I can be sure, it didn't iterate once fallen into audience evaluation but didn't bucket.

Reply:
Since there are 3 rollout rules `experiments` in this test, we have only mocked response for `Evaluate` and `GetDecision` methods of 1st and last rule, This makes sure that if these methods were called for any other experiments, the test will fail. We assert that the returned rollout rule was the last one already by providing `testExp1118 ` in expectedFeatureDecision.